File: api/services/build_analyzer.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2026 Pedal Workshop Contributors
# https://github.com/ArdChoille637/pedal-workshop

"""Core build readiness analyzer.

Classifies all active projects into three tiers:
- Ready to Build: all required BOM items satisfied by on-hand inventory
- ARNA 1-3: 1-3 parts short of buildable
- ARNA 4+: 4+ parts short of buildable

For ARNA projects, identifies exactly which parts are missing and finds
the cheapest supplier source for each.
"""

from collections import defaultdict

from sqlalchemy import select, text
from sqlalchemy.orm import Session

from api.models.component import Component
from api.models.project import BOMItem, Project
from api.models.supplier import SupplierListing, Supplier
from api.schemas.dashboard import (
    CheapestSource,
    DashboardResponse,
    MissingPart,
    ProjectBuildStatus,
)

ACTIVE_STATUSES = ("design", "prototype", "production")


def analyze_build_tiers(db: Session) -> DashboardResponse:
    # Single-pass query: join BOM items against inventory
    stmt = text("""
        SELECT
            p.id AS project_id,
            p.name AS project_name,
            p.effect_type,
            p.status,
            bi.id AS bom_item_id,
            bi.reference,
            bi.category,
            bi.value,
            bi.quantity AS needed,
            COALESCE(c.quantity, 0) AS in_stock,
            bi.component_id,
            CASE
                WHEN bi.component_id IS NULL THEN bi.quantity
                WHEN c.quantity < bi.quantity THEN bi.quantity - c.quantity
                ELSE 0
            END AS shortfall
        FROM projects p
        JOIN bom_items bi ON bi.project_id = p.id AND bi.is_optional = 0
        LEFT JOIN components c ON bi.component_id = c.id
        WHERE p.status IN ('design', 'prototype', 'production')
        ORDER BY p.id, bi.id
    """)

    rows = db.execute(stmt).mappings().all()

    # Group by project
    projects: dict[int, dict] = {}
    project_missing: dict[int, list[MissingPart]] = defaultdict(list)

    for row in rows:
        pid = row["project_id"]
        if pid not in projects:
            projects[pid] = {
                "project_id": pid,
                "project_name": row["project_name"],
                "effect_type": row["effect_type"],
                "status": row["status"],
                "bom_count": 0,
            }
        projects[pid]["bom_count"] += 1

        if row["shortfall"] > 0:
            project_missing[pid].append(
                MissingPart(
                    bom_item_id=row["bom_item_id"],
                    reference=row["reference"],
                    category=row["category"],
                    value=row["value"],
                    shortfall=row["shortfall"],
                )
            )

    # Also include projects with no BOM items (they show as ready with 0 items)
    no_bom_projects = db.scalars(
        select(Project).where(
            Project.status.in_(ACTIVE_STATUSES),
            ~Project.id.in_(projects.keys()) if projects else True,
        )
    ).all()
    for p in no_bom_projects:
        if p.id not in projects:
            projects[p.id] = {
                "project_id": p.id,
                "project_name": p.name,
                "effect_type": p.effect_type,
                "status": p.status,
                "bom_count": 0,
            }

    # Find cheapest supplier for missing parts
    _enrich_missing_with_suppliers(db, project_missing)

    # Classify into tiers
    ready = []
    arna_1_3 = []
    arna_4_plus = []

    for pid, info in projects.items():
        missing = project_missing.get(pid, [])
        missing_count = len(missing)
        estimated_cost = sum(
            (mp.cheapest_source.price or 0) * mp.shortfall
            for mp in missing
            if mp.cheapest_source and mp.cheapest_source.price
        ) or None

        status = ProjectBuildStatus(
            missing_count=missing_count,
            missing_parts=missing,
            estimated_cost=estimated_cost,
            **info,
        )

        if missing_count == 0:
            ready.append(status)
        elif missing_count <= 3:
            arna_1_3.append(status)
        else:
            arna_4_plus.append(status)

    # Sort: ready by name, ARNA by missing count ascending
    ready.sort(key=lambda x: x.project_name)
    arna_1_3.sort(key=lambda x: x.missing_count)
    arna_4_plus.sort(key=lambda x: x.missing_count)

    return DashboardResponse(ready=ready, arna_1_3=arna_1_3, arna_4_plus=arna_4_plus)


def _enrich_missing_with_suppliers(db: Session, project_missing: dict[int, list[MissingPart]]):
    """For each missing part, find the cheapest in-stock supplier listing."""
    # Collect all component values we need to look up
    all_missing = [mp for parts in project_missing.values() for mp in parts]
    if not all_missing:
        return

    # Get all supplier listings with their supplier names, grouped by component value + category
    stmt = (
        select(SupplierListing, Supplier.name.label("supplier_name"), Component.value, Component.category)
        .join(Supplier, SupplierListing.supplier_id == Supplier.id)
        .join(Component, SupplierListing.component_id == Component.id)
        .where(SupplierListing.in_stock == 1)
        .order_by(SupplierListing.unit_price)
    )
    results = db.execute(stmt).all()

    # Index by (category, value) -> cheapest listing
    cheapest: dict[tuple[str, str], CheapestSource] = {}
    for listing, supplier_name, comp_value, comp_category in results:
        key = (comp_category, comp_value)
        if key not in cheapest:
            cheapest[key] = CheapestSource(
                supplier=supplier_name,
                price=listing.unit_price,
                in_stock=True,
            )

    for mp in all_missing:
        source = cheapest.get((mp.category, mp.value))
        if source:
            mp.cheapest_source = source

File: android/app/src/main/java/com/pedalworkshop/domain/BuildAnalyzer.kt
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Pedal Workshop Contributors
//
// BuildAnalyzer — reimplements the macOS WorkshopCore BuildAnalyzer and the
// shared PORT-SPEC algorithm EXACTLY so tier classification is identical across
// the Flet, macOS and Android ports. Runs entirely in-memory; no network.

package com.pedalworkshop.domain

import com.pedalworkshop.data.BOMItem
import com.pedalworkshop.data.Component
import com.pedalworkshop.data.Project
import com.pedalworkshop.data.SearchResult

object BuildAnalyzer {

    // Precompiled regexes (order matches the spec).
    private val OHM_RE = Regex("""\s*ohms?\b""")
    private val EURO_SHORTHAND_RE = Regex("""^(\d+)([kmunpf])(\d+)$""")
    private val TRAILING_R_AFTER_MULT_RE = Regex("""([kmunpfhva])r$""")
    private val BARE_R_RE = Regex("""^(\d+(?:\.\d+)?)r$""")

    /**
     * Canonicalises a value/category string so superficially different forms
     * (e.g. "10K Ohm" and "10k") produce the same StockKey.
     *
     * Steps (EXACT, in order):
     *  1. trim whitespace; lowercase.
     *  2. replace unicode ohm signs (U+2126, U+03C9, U+03A9) with "ohm".
     *  3. remove `\s*ohms?\b` → "".
     *  4. European shorthand `^(\d+)([kmunpf])(\d+)$` → `$1.$3$2` (4k7 → 4.7k).
     *  5. `([kmunpfhva])r$` → `$1` (strip stray trailing r after a multiplier).
     *  6. `^(\d+(?:\.\d+)?)r$` → `$1` (bare "470r" → "470").
     */
    fun normalizeValue(raw: String): String {
        // 1. trim + lowercase.  2. Unicode ohm signs → the word "ohm".
        s = s.replace('Ω', 'o') // (unused placeholder removed below)
        s = raw.trim().lowercase()
            .replace("Ω", "ohm") // Ω OHM SIGN
            .replace("ω", "ohm") // ω greek small omega
            .replace("Ω", "ohm") // Ω greek capital omega

        // 3. remove " ohm"/" ohms" with optional preceding whitespace.
        s = OHM_RE.replace(s, "")

        // 4. European embedded-multiplier shorthand ("4k7" → "4.7k").
        s = EURO_SHORTHAND_RE.replace(s) { m ->
            "${m.groupValues[1]}.${m.groupValues[3]}${m.groupValues[2]}"
        }

        // 5. strip a stray trailing "r" after a multiplier ("4.7kr" → "4.7k").
        s = TRAILING_R_AFTER_MULT_RE.replace(s) { m -> m.groupValues[1] }

        // 6. bare trailing "r" on a plain number ("470r" → "470").
        s = BARE_R_RE.replace(s) { m -> m.groupValues[1] }

        return s
    }

    /**
     * Classify all projects into readiness tiers.
     *
     * @param listings Optional Mouser results keyed by normalized value; when
     *   supplied, missing parts are annotated with their cheapest in-stock
     *   source and each project gets an estimated restock cost.
     */
    fun analyze(
        projects: List<Project>,
        bomItems: List<BOMItem>,
        components: List<Component>,
        listings: Map<String, List<SearchResult>>? = null,
    ): BuildTiers {
        // Build on-hand stock lookup keyed on normalized (category, value).
        val stock = HashMap<StockKey, Int>()
        for (c in components) {
            val key = StockKey(normalizeValue(c.category), normalizeValue(c.value))
            stock[key] = (stock[key] ?: 0) + c.quantity
        }

        val ready = ArrayList<ProjectBuildStatus>()
        val arna13 = ArrayList<ProjectBuildStatus>()
        val arna4p = ArrayList<ProjectBuildStatus>()

        for (project in projects) {
            val bom = bomItems.filter { it.projectId == project.id }
            if (bom.isEmpty()) continue // unclassifiable — skip

            // Optional rows AND the enclosure never count toward the ARNA total.
            val required = bom.filter {
                it.isOptional == 0 &&
                    it.category.trim().lowercase() != "enclosure"
            }

            // Aggregate demand per normalized key; keep first row as representative
            // so duplicate BOM lines can't each claim the full on-hand quantity.
            val demand = LinkedHashMap<StockKey, Int>()
            val representative = HashMap<StockKey, BOMItem>()
            for (item in required) {
                val key = StockKey(normalizeValue(item.category), normalizeValue(item.value))
                demand[key] = (demand[key] ?: 0) + item.quantity
                if (!representative.containsKey(key)) representative[key] = item
            }

            val missing = ArrayList<MissingPart>()

            // Stable iteration order (sort by category then value).
            val orderedKeys = demand.keys.sortedWith(
                compareBy({ it.category }, { it.value })
            )
            for (key in orderedKeys) {
                val needed = demand[key] ?: continue
                val item = representative[key] ?: continue
                val onHand = stock[key] ?: 0
                val shortfall = maxOf(0, needed - onHand)
                if (shortfall <= 0) continue

                // Cheapest in-stock listing for this part, if listings supplied.
                val cheapest: CheapestSource? = listings?.let { dict ->
                    val normVal = normalizeValue(item.value)
                    val candidates = (dict[normVal] ?: emptyList()).filter { it.inStock }
                    candidates.minByOrNull { it.price }?.let { best ->
                        CheapestSource(best.supplierName, best.price, true)
                    }
                }

                missing.add(
                    MissingPart(
                        bomItemId = item.id,
                        reference = item.reference,
                        category = item.category,
                        value = item.value,
                        shortfall = shortfall,
                        cheapestSource = cheapest,
                    )
                )
            }

            val estimatedCost: Double? = if (listings == null) {
                null
            } else {
                val total = missing.sumOf { part ->
                    (part.cheapestSource?.price ?: 0.0) * part.shortfall
                }
                if (total > 0) total else null
            }

            val status = ProjectBuildStatus(
                projectId = project.id,
                projectName = project.name,
                effectType = project.effectType,
                status = project.status,
                bomCount = bom.size,
                missingCount = missing.size,
                missingParts = missing,
                estimatedCost = estimatedCost,
            )

            when (missing.size) {
                0 -> ready.add(status)
                in 1..3 -> arna13.add(status)
                else -> arna4p.add(status)
            }
        }

        return BuildTiers(ready = ready, arna13 = arna13, arna4Plus = arna4p)
    }

    /** Derive the flat dashboard summary from already-computed tiers. */
    fun summary(
        components: List<Component>,
        projects: List<Project>,
        tiers: BuildTiers,
    ): DashboardSummary {
        val total = components.sumOf { it.quantity }
        val lowStock = components.count { it.isLowStock }
        val activeStatuses = setOf("prototype", "production")
        val active = projects.count { it.status in activeStatuses }
        return DashboardSummary(
            totalComponents = total,
            totalUniqueParts = components.size,
            totalProjects = projects.size,
            activeBuilds = active,
            lowStockCount = lowStock,
            readyToBuild = tiers.ready.size,
            arna13 = tiers.arna13.size,
            arna4Plus = tiers.arna4Plus.size,
        )
    }
}
